p2p/pex/addrbook_test.go

```diff
@@ -36,6 +38,202 @@ func deleteTempFile(fname string) {
+}
+
+// Creates a Book with m old and n new addresses.
+func createAddrBookWithMOldAndNNewAddrs(t *testing.T, m, n int) *addrBook {
+	fname := createTempFileName("addrbook_test")
+	defer deleteTempFile(fname)
+
+	book := NewAddrBook(fname, true)
+	book.SetLogger(log.TestingLogger())
+	assert.Zero(t, book.Size())
+
+	// add m old addresses
+	randAddrs := randNetAddressPairs(t, m)
+	for _, addr := range randAddrs {
+		book.AddAddress(addr.addr, addr.src)
+		book.MarkGood(addr.addr)
+	}
+	// add n new addresses
+	randAddrs = randNetAddressPairs(t, n)
+	for _, addr := range randAddrs {
+		book.AddAddress(addr.addr, addr.src)
+	}
+
+	return book
+}
+
+// Counts the number of Old and New addresses
+func countOldAndNewInAddrs(addrs []*p2p.NetAddress, book *addrBook) (nOld, nNew int) {
+
+	for _, addr := range addrs {
+		if addr == nil {
+			continue
+		}
+		if book.IsGood(addr) {
+			nOld++
+		} else {
+			nNew++
+		}
+	}
+	return nOld, nNew
+}
+
+// Analyse the layout of the selection specified by 'addrs'
+// Returns:
+// - seqLens - the lengths of the sequences of addresses of same type
+// - seqTypes - the types of sequences in selection
+func analyseSelectionLayout(book *addrBook, addrs []*p2p.NetAddress) (seqLens, seqTypes []int, err error) {
+	// address types are: 0 - nil, 1 - new, 2 - old
+	prevt := 0
+
+	cnt := 0
+	for i, addr := range addrs {
+		if addr == nil {
+			err = fmt.Errorf("nil addresses in selection, index %d", i)
+			return nil, nil, err
+		}
+		addrt := 0
+		if book.IsGood(addr) {
+			addrt = 2
+		} else {
+			addrt = 1
+		}
+
+		if addrt != prevt && prevt != 0 {
+			seqLens = append(seqLens, cnt)
+			seqTypes = append(seqTypes, prevt)
+			cnt = 0
+		}
+		cnt++
+		prevt = addrt
+	}
+	seqLens = append(seqLens, cnt)
+	seqTypes = append(seqTypes, prevt)
+	return seqLens, seqTypes, nil
+}
+
+func TestEmptyAddrBookSelection(t *testing.T) {
+	book := createAddrBookWithMOldAndNNewAddrs(t, 0, 0)
+	addrs := book.GetSelectionWithBias(biasToSelectNewPeers)
+	assert.Nil(t, addrs, "expected a nill selection")
+}
+
+func testAddrBookAddressSelection(t *testing.T, b int) {
+	// generate all combinations of old (m) and new (n=b-m) addresses
+	for m := 0; m < b+1; m++ {
+		n := b - m
+		// debug string to show the context for potential failures
+		dbgStr := fmt.Sprintf("book: size %d, number new % d, old %d", b, n, m)
+
+		// create book and get selection
+		book := createAddrBookWithMOldAndNNewAddrs(t, m, n)
+		addrs := book.GetSelectionWithBias(biasToSelectNewPeers)
+
+		r := len(addrs)
+		assert.NotNil(t, addrs, "%s - expected a non-nill selection", dbgStr)
+		assert.NotZero(t, r, "%s - expected at least one address in selection", dbgStr)
+
+		// verify that the number of old and new addresses are the ones expected
+		numOld, numNew := countOldAndNewInAddrs(addrs, book)
+		assert.Equal(t, r, numOld+numNew, "%s - expected selection completely filled", dbgStr)
+
+		// Given:
+		// n - num new addrs, m - num old addrs
```

Comment:
why use one letter vars when you can just use numNewAddrs or numNew => developer won't be forced to seek the Given reference to understand the code

i=min(n, k) => `numNewInTheBeginning`??